### DISCUSSION TITLE
Add explicit type hints for public properties in `Response`

### DISCUSSION THREAD

**Discussion by RobertCraigie** [encode/httpx#2964] [Potential Issue]
Mypy will report an error for this code:

```py
from typing_extensions import reveal_type
from httpx import Response

r = Response(status_code=200)
assert r.is_stream_consumed is False
r.read()
assert r.is_stream_consumed is True
reveal_type(r.is_stream_consumed)  # error: unreachable
```

This is because mypy's inference algorithm doesn't properly handle the `is_stream_consumed` assignment and thinks that `r.is_stream_consumed` can only ever be `False`. While this is also completely a mypy bug, imo it is always helpful to explicitly declare the types for class properties and in this case would obviate the mypy bug.

I'd happily open a PR for this if the change would be acceptable.

**Comment by lovelydinosaur** [comment]
I'm not sure...

```python
>>> from typing_extensions import reveal_type
>>> from httpx import Response
>>> 
>>> r = Response(status_code=200)
>>> assert r.is_stream_consumed is False
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError
```

Are you sure you're not seeing the "error: unreachable" because there's an incorrect assertion there?

**Reply by RobertCraigie** [reply]
Sorry that was example code that I hadn't ever actually ran at runtime, just to demonstrate the underlying static type issue.

I encountered this in a scenario where I was actually making requests and the tests passed at runtime.
